ibis/backends/tests/test_client.py

```diff
@@ -626,6 +651,98 @@ def _emp(a, b, c, d):
     assert len(con.table(employee_data_1_temp_table).execute()) == 3
 
 
+@NO_MERGE_SUPPORT
+def test_upsert_from_dataframe(
+    backend, con, employee_data_1_temp_table, test_employee_data_3
+):
+    temporary = con.table(employee_data_1_temp_table)
+    df1 = temporary.execute().set_index("first_name")
+
+    con.upsert(employee_data_1_temp_table, obj=test_employee_data_3, on="first_name")
+    result = temporary.execute()
+    df2 = test_employee_data_3.set_index("first_name")
+    expected = pd.concat([df1[~df1.index.isin(df2.index)], df2]).reset_index()
+    assert len(result) == len(expected)
+    backend.assert_frame_equal(
+        result.sort_values("first_name").reset_index(drop=True),
+        expected.sort_values("first_name").reset_index(drop=True),
+    )
+
+
+@NO_MERGE_SUPPORT
+@pytest.mark.parametrize("with_order_by", [True, False])
+def test_upsert_from_expr(
+    backend, con, employee_data_1_temp_table, employee_data_3_temp_table, with_order_by
+):
+    temporary = con.table(employee_data_1_temp_table)
+    from_table = con.table(employee_data_3_temp_table)
+    if with_order_by:
+        if backend.name() == "mssql":
+            pytest.xfail(
+                "MSSQL doesn't allow ORDER BY in subqueries, unless "
+                "TOP, OFFSET or FOR XML is also specified"
+            )
```

Comment:
```suggestion
```

Instead of having the `xfail` inline, can you add it as a parameter mark?

Instead of `True`, above on line 673, you can do something like `pytest.param(True, marks=pytest.mark.notyet("mssql"))`